### ISSUE TITLE
CW_L2 attack did not work as expected or expressed

### ISSUE THREAD

**Issue by persistz** [bethgelab/foolbox#387]
One parameter of CW_l2 attack is confidence, which determines the robust or probability of adversarial examples they generated.
But in FoolBox, it seems no difference between I set 'Confidence=0' and 'Confidence=30'.
In CW's git repo, they only confirm the adversarial is an adversarial example, until the max_prob-second_prob>confidence, but I did not find this code in Foolbox.

**Comment by persistz** [comment]
Sorry for the poor format, update the code again:
In Carlini's codes:

`def compare(x,y):
if not isinstance(x, (float, int, np.int64)):
x = np.copy(x)
if self.TARGETED:
x[y] -= self.CONFIDENCE
else:
x[y] += self.CONFIDENCE
x = np.argmax(x)
if self.TARGETED:
return x == y
else:
return x != y`

In Foolbox:

 `# is_adv is True as soon as the is_adv_loss goes below 0
 # but sometimes we want additional confidence
    is_adv_loss += confidence
    is_adv_loss = max(0, is_adv_loss)`

But...I cannot find the code to support "is_adv is True as soon as the is_adv_loss goes below 0".
I only see:
logits, is_adv = yield from a.forward_one(x)

**Comment by jonasrauber** [comment]
Good point! I guess the problem is that the `criterion` does not reflect the desire for a _confident_ misclassification. You could try also specifying this confidence as part of the `criterion` that you pass to the attack. Foolbox does provide `foolbox.criteria.ConfidentMisclassification` but it has a somewhat different meaning (let's you specify the probability to achieve).
The right thing is probably to add a criterion that matches the definition of _confidence_ used by the CW attack and (automatically) using this criterion in the CW attack if confidence is set.
Could you give this a try?

**Comment by persistz** [comment]
Yes, I have known that 'foolbox.criteria.ConfidentMisclassification', I generated lots of high confidence adversarial examples with this criterion and BIM.

What I want to focus more on this question is whether we can get the implementation consistent with the original C&W attack method through simple modification.

For example, could we change:
```
logits, is_adv = yield from a.forward_one(x)
```
to 
```
logits, _ = yield from a.forward_one(x)
is_adv = 1 if logits[adv]-logits[second_higher] > confidence else 0
```

This is just a small demo, I am not sure if this method works.
But I think this may be a better solution for this kind of method, as if we use ConfidentMisclassification as a criterion, we have to change the parameters defination of C&W. What do you think?

**Comment by jonasrauber** [comment]
I don't think that will work, because Foolbox always returns the adversarial (according to the gieven `criterion`) with the smallest perturbation (according to the given `distance`).

**Comment by persistz** [comment]
Got it. 👍 I will try to make some progress. And I think if the member team could modify the documentation for this part will be better which could alleviate the doubts of other users.
I will close this question.

**Comment by jonasrauber** [comment]
@persistz Foolbox is an open source project. It would be great if you were to contribute this improvement to the documentation.
